### PR TITLE
Where tuple return

### PR DESCRIPTION
**Pull request by nfarabullini**
Implementation of tuple return in where with either fields or scalars in returns, e.g.
```python
@field_operator
    def conditional_tuple_with_fields(
        mask: Field[[IDim], bool], a: Field[[IDim], float64], b: Field[[IDim], float64]
    ) -> tuple[Field[[IDim], float64], Field[[IDim], float64]]:
        return where(mask, (a, b), (b, a))
```
 
or

```python
@field_operator
    def conditional_tuple_with_fields_and_scalars(
        mask: Field[[IDim], bool], a: Field[[IDim], float64], b: Field[[IDim], float64]
    ) -> tuple[Field[[IDim], float64], Field[[IDim], float64]]:
        return where(mask, (a, b), (1.0, 2.0))
```

as well as nested tuples

### CODE REVIEW

File: src/functional/ffront/type_info.py
Comment (DropD) [inline src/functional/ffront/type_info.py:388]: Since the semantic meaning of this function and the terms used in it's code and docstring are quite limited to the only place it is being used, it should probably be moved to `type_deduction.py` until it turns out to be more widely useful.

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:511]: I know this touches existing code, but I would still suggest changing `left` and `right` to `true_branch` and `false_branch` or similar. It is quite a contained change but a big plus for readability. 

For operators it makes sense to use `left` and `right` because there is no semantic difference between the left and right operand but here we can call things by their meaning.

**Review by DropD — CHANGES_REQUESTED** [review]
Currently this seems to be limited to flat tuples (no nesting). At the very least there is no test that demonstrates otherwise. Is that intentional? Similar things like tuple returns for field operators allow nested tuples.

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:582]: Since `_extract_promoted_tuple_type`, `_construct_tuple_type`, `_promote_to_mask_type` deal with types and none of them makes use of the `self` instance, except to call themselves, I would suggest making them free functions. The precedent for this is the `boolified_type` function in the same file.

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:552]: It seems a bit unintuitive that this function should return on the first `TupleType` element, completely discarding all succeeding elements. Is that intentional?

With the current nested tuple test it doesn't seem to make a difference either way, that's why I ask.

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:560]: If this condition could be built into `promote_to_mask_type` instead (or extracted into another function), then the whole `for` level `else` branch could be written

```python
            else:
                return_type = type_info.promote(return_type, element) if return_type else element
                return_type = self._promote_to_mask_type(mask_tye, return_type)  # or new function if preferred
```

Which significantly reduces the brain power required to follow the control flow (for me at least).

**Review by DropD — CHANGES_REQUESTED** [review]

**Review by DropD — APPROVED** [review]
Looks good, thank you! I'll leave it up to you if you want to add docstrings to the type handling functions.

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:542]: Actually, I just realized there is no test for any of this in `test_type_deduction`. 

The problem with this is that the types may be deduced completely wrongly here while all the tests still pass. Let's fix that before merging.

**Review by DropD — CHANGES_REQUESTED** [review]

File: tests/functional_tests/ffront_tests/test_type_deduction.py
Comment (DropD) [inline tests/functional_tests/ffront_tests/test_type_deduction.py:536]: The following **crucial** test is missing:

```python
ADim = Dimension("ADim")
BDim = Dimension("BDim")

def tuple_where(a: Field[[ADim], bool], b: Field[[ADim], float64], c: Field[[ADim, BDim], float64]):
    return where(a, ((c, 9.0), (b, 6.0)), ((8.0, b), (5.0, 9.0)))
```

I am not 100 % sure what the return type should be but I would expect something like

```python
(
    (Field[[ADim, BDim], ...], Field[[ADim], ...]),
    (Field[[ADim], ...], Field[[ADim], ...]),
)
```

In any case looking at the code I don't think the question this test asks has been addressed: "What happens if I mix dimensions in a nested tuple (in a way that the two branches are compatible)?".

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:58]:
```suggestion
    Extract argument type in tuple to which all others need to be promoted.

    If arguments types in a tuple are different (e.g. some are Fields and others Scalars), they all need to be promoted to be the same.
    This function extracts the type with respect to which the promotion has to be done.
```

In addition to the phrasing I am questioning the described semantics. Do all tuple elements necessarily have to be the same? What happens if the tuples contain mixed dimension fields? See comment in the type deduction tests for a better description of the question.

File: tests/functional_tests/ffront_tests/test_type_deduction.py
Comment (nfarabullini) [reply tests/functional_tests/ffront_tests/test_type_deduction.py:536]: As the implementation is atm, the return dimensions should be [ADim, BDim] for all tuples items.

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (nfarabullini) [reply src/functional/ffront/foast_passes/type_deduction.py]: If you mix dimensions, all items will be promoted to a mutual type. Such type contains all dimensions from all inout fields

File: tests/functional_tests/ffront_tests/test_type_deduction.py
Comment (havogt) [reply tests/functional_tests/ffront_tests/test_type_deduction.py:536]: Yes, I think we should discuss this case together.

File: src/functional/iterator/embedded.py
Comment (havogt) [inline src/functional/iterator/embedded.py:846]: I think the changes here should be rolled back, see my other comment.

File: tests/functional_tests/ffront_tests/test_execution.py
Comment (havogt) [inline tests/functional_tests/ffront_tests/test_execution.py:838]: The reason you had to touch embedded is not because of how where behave, but because of your test setup here. All out fields should have the same dimensions.

One option would be to take the result of the where and add all the tuple elements into a single out field if you want to keep that test. But I don't think this tests is actually testing something extra that is not already tested (maybe I am missing the point).

**Review by havogt — CHANGES_REQUESTED** [review]
Just reviewed the changes in embedded part. Passing further review back to @DropD .

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:55]:
```suggestion
    Recursively construct  the return types for the tuple return branch.
```

File: src/functional/ffront/foast_passes/type_deduction.py
Comment (DropD) [inline src/functional/ffront/foast_passes/type_deduction.py:50]: Is there any case where `element_types` does not refer to the types in the "true-branch"? If not, would you agree that there would be some benefit in renaming?

Comment (nfarabullini) [reply src/functional/ffront/foast_passes/type_deduction.py]: agree and edited

File: tests/functional_tests/ffront_tests/test_execution.py
Comment (DropD) [inline tests/functional_tests/ffront_tests/test_execution.py:765]: shouldn't the result be checked here?

**Review by DropD — APPROVED** [review]
There is one test where the check of the result seems to be missing, however I consider it non-critical, since it seems to be more or less covered by other tests.

File: tests/functional_tests/ffront_tests/test_execution.py
Comment (havogt) [reply tests/functional_tests/ffront_tests/test_execution.py:765]: Just for clarity: if the test is useful, but the result of execution is not important, the test should be a parsing test or lowering test, but not in `test_execution`.

File: src/functional/ffront/past_to_itir.py
Comment (havogt) [reply src/functional/ffront/past_to_itir.py:245]: this was a pending comment that I apparently never submitted

**Review by havogt — CHANGES_REQUESTED** [review]

**Review by havogt — APPROVED** [review]

**Pull request merged**
